sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/nodes/intrinsics/multithreading/UtilFunctionCall.java

```diff
@@ -0,0 +1,178 @@
+            } finally {
+                // call destructors from key create
+                if (this.isThread) {
+                    for (int key = 1; key <= ctx.curKeyVal; key++) {
+                        LLVMPointer destructor = UtilAccessCollectionWithBoundary.get(ctx.destructorStorage, key);
+                        if (destructor != null && !destructor.isNull()) {
+                            ConcurrentMap<Long, LLVMPointer> specValueMap = UtilAccessCollectionWithBoundary.get(ctx.keyStorage, key);
+                            // if key was deleted continue with next destructor
+                            if (specValueMap == null) {
+                                continue;
+                            }
+                            Object keyVal = UtilAccessCollectionWithBoundary.get(specValueMap, Thread.currentThread().getId());
+                            if (keyVal != null) {
+                                // if key value is null pointer continue with next destructor
+                                try {
+                                    LLVMPointer keyValPointer = LLVMPointer.cast(keyVal);
+                                    if (keyValPointer.isNull()) {
+                                        continue;
+                                    }
+                                } catch (Exception e) {
+                                    // ignored
+                                }
+                                UtilAccessCollectionWithBoundary.remove(specValueMap, Thread.currentThread().getId());
+                                new FunctionCallRunnable(destructor, keyVal, this.ctx, false).run();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    static final class MyArgNode extends LLVMExpressionNode {
+
+        private final FrameSlot slot;
+
+        private MyArgNode(FrameSlot slot) {
+            this.slot = slot;
+        }
+
+        @Override
+        public Object executeGeneric(VirtualFrame frame) {
+            return frame.getValue(slot);
+        }
+    }
+
+    private static final class FunctionCallNode extends RootNode {
+
+        @Child LLVMExpressionNode callNode = null;
+
+        @CompilationFinal FrameSlot functionSlot = null;
+
+        @CompilationFinal FrameSlot argSlot = null;
+
+        @CompilationFinal FrameSlot spSlot = null;
+
+        private final LLVMContext ctx;
+
+        FunctionCallNode(LLVMLanguage language) {
+            super(language);
+            this.ctx = language.getContextReference().get();
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            LLVMStack.StackPointer sp = ctx.getThreadingStack().getStack().newFrame();
+            if (callNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                functionSlot = frame.getFrameDescriptor().findOrAddFrameSlot("function");
```

Comment:
You can create a `FrameDescriptor` and these frameslots yourself in the constructor. Have a look at the `RootNode` constructor that takes such a manually defined descriptor.

Reply:
ok, added all the slots before initializing the function call node, and then passed the descriptor in the root node constructor.
from then on in the node class used `this.getFrameDescriptor()` and the already initialized frames.

Reply:
What do you need `this.getFrameDescriptor()` for? If you add the frameslots to the descriptor in the constructor already, you can make the slots in the node `final` and have no use for the descriptor at runtime / outside the constructor anymore.

P.S. Please also annotate the constructor with `@TruffleBoundary`.

Reply:
i was not sure if there would maybe be some internal changes to these slots, so i always took the "current" slot from the frame descriptor.
i now set the final slots in the constructor from the descriptor find slot function, and then in the execute function just use these final slot fields.
i hope that's what you meant.

Reply:
Yes, that's what I meant. Once a slot is added to the frame, the `FrameSlot` object is valid until the slot is explicitly removed again, so you don't need to keep polling the descriptor for a new one.